### ISSUE TITLE
[no-non-null-assertion] Rule proposal:  no-non-null-assertion allow assertion with comment

### ISSUE THREAD

**Issue by xiaoxiangmoe** [typescript-eslint/typescript-eslint#414]
Examples of incorrect code for this rule:

```ts
interface Foo {
  bar?: string;
}

const foo: Foo = getFoo()!;
const includesBaz: boolean = foo.bar!.includes('baz');
```
 

Examples of correct code for this rule:

```ts
interface Foo {
  bar?: string;
}

// Assertion: getFoo() always return non-null Foo
const foo: Foo = getFoo()!;
const bar = foo2.bar!; // Assertion: getFoo() always return Foo's bar as non-null string
const includesBaz: boolean = bar.includes('baz');
```

---

If there exists an comment start with `Assertion:`, allow use non null assertion.

**Comment by bradzacher** [comment]
You can achieve this via eslint comments: https://eslint.org/docs/user-guide/configuring#disabling-rules-with-inline-comments

It's a lot more verbose, of course, but it provides the same functionality for free:

```ts
interface Foo {
  bar?: string;
}

/* eslint-disable-next-line @typescript-eslint/no-non-null-assertion */// getFoo() always return non-null Foo
const foo: Foo = getFoo()!;
const bar = foo2.bar!; /* eslint-disable-line @typescript-eslint/no-non-null-assertion */// getFoo() always return Foo's bar as non-null string
const includesBaz: boolean = bar.includes('baz');
```

I'm against the idea of building in our own comment infrastructure for disabling a rule when eslint provides the functinoality natively
